### ISSUE TITLE
TestPickleDataset fails on Windows

### ISSUE THREAD

**Issue by disktnk** [chainer/chainer#7768]
- Chainer version: master, v6
- OS/Platform: Windows

```
C:\path\chainer>pytest tests\chainer_tests\datasets_tests\test_pickle_dataset.py::TestPickleDataset
================================================= test session starts =================================================
platform win32 -- Python 3.7.3, pytest-4.1.1, py-1.8.0, pluggy-0.12.0
rootdir: C:\path\chainer, inifile: setup.cfg
collected 2 items

tests\chainer_tests\datasets_tests\test_pickle_dataset.py F.                                                     [100%]

====================================================== FAILURES =======================================================
__________________________________________ TestPickleDataset.test_after_fork __________________________________________

self = <chainer_tests.datasets_tests.test_pickle_dataset.TestPickleDataset testMethod=test_after_fork>

    def test_after_fork(self):
        writer = datasets.PickleDatasetWriter(self.io)
        writer.write(1)
        writer.flush()

        reader = ReaderMock(self.io)
        # Assign to avoid destruction of the instance
        # before creation a child process
        dataset = datasets.PickleDataset(reader)

        assert reader.n_hook_called == 0
        p = multiprocessing.Process()
        p.start()
        p.join()
>       assert reader.n_hook_called == 1
E       AssertionError: assert 0 == 1
E        +  where 0 = <chainer_tests.datasets_tests.test_pickle_dataset.ReaderMock object at 0x000001E3315518D0>.n_hook_called

tests\chainer_tests\datasets_tests\test_pickle_dataset.py:75: AssertionError
```

https://github.com/chainer/chainer/blob/09252fb964a7159156b3975033c46835717841ae/chainer/datasets/pickle_dataset.py#L101
`register_after_fork` does not work? (I don't know detail)

**Comment by zaltoprofen** [comment]
Oh, I had forgot `fork()` system call is not available on windows...
cf. https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods

Additionally, my implementation (#7625) is not probably enough to make available multiprocessing with `spawn` method.
It seems that we need to add `__getstate__` and` __setstate__` methods into `_FileReader` and reopen the file in it.

**Comment by toslunar** [comment]
Fixed by #8465.
